src/vmspawn/vmspawn-util.c

```diff
@@ -236,3 +266,75 @@ int find_qemu_binary(char **ret_qemu_binary) {
 
         return find_executable(qemu_arch_specific, ret_qemu_binary);
 }
+
+int vsock_fix_child_cid(unsigned *machine_cid, const char *machine, int *ret_child_sock) {
+        _cleanup_close_ int vfd = -EBADF;
+        int r, i;
+        static const uint8_t sip_key[HASH_KEY_SIZE] = { 0 };
+        struct siphash machine_hash_state, state;
+
+        /* uint64_t is required here for the ioctl call, but valid CIDs are only 32 bits */
+        uint64_t cid = *ASSERT_PTR(machine_cid);
+
+        assert(machine);
+        assert(ret_child_sock);
+
+        /* Fix the CID of the AF_VSOCK socket passed to qemu
+         *
+         * If the user has passed us a CID (machine_cid != VMADDR_CID_ANY), then attempt to bind to that CID
+         * and error if we cannot.
+         *
+         * Otherwise hash the machine name to get a random CID and attempt to bind to that.
+         * If it is occupied add more information into the hash and try again.
+         * If after 1000 attempts this hasn't worked fallback to truly random CIDs.
+         * If after another 1000 attempts this hasn't worked then give up and return EADDRNOTAVAIL.
+         */
+
+        /* don't open with O_CLOEXEC, this fd is passed to QEMU */
+        vfd = open("/dev/vhost-vsock", O_RDWR);
+        if (vfd < 0)
+                return log_debug_errno(errno, "Failed to open /dev/vhost-vsock as read/write: %m");
+
+        if (cid != VMADDR_CID_ANY) {
+                r = ioctl(vfd, VHOST_VSOCK_SET_GUEST_CID, &cid);
+                if (r < 0)
+                        return log_debug_errno(errno, "Failed to set CID for child vsock with user provided CID %lu: %m", cid);
+                *ret_child_sock = TAKE_FD(vfd);
+                return 0;
+        }
+
+        siphash24_init(&machine_hash_state, sip_key);
+        siphash24_compress_string(machine, &machine_hash_state);
+        for (i = 0; i < 1000; i++) {
+                state = machine_hash_state;
+                uint8_t i_bytes[sizeof i];
+                memcpy(i_bytes, &i, sizeof i);
```

Comment:
why copy this? you can just pass a direct pointer to `i` to siphash24_compress_safe()

Reply:
Fixed in ef3bb39e80.